timelock-server/src/main/java/com/palantir/atlasdb/timelock/clock/ClockSkewMonitor.java

```diff
@@ -0,0 +1,176 @@
+
+    private final Logger log = LoggerFactory.getLogger(ClockSkewMonitor.class);
+    private final Map<String, ClockService> monitorsByServer;
+    private final Map<String, RequestTime> previousRequestsByServer;
+    private final Supplier<Boolean> shouldRunClockSkewMonitor;
+
+    public static ClockSkewMonitor create(Set<String> remoteServers, Optional<SSLSocketFactory> optionalSecurity,
+            Supplier<Boolean> shouldRunClockSkewMonitor) {
+        Map<String, ClockService> monitors = Maps.toMap(remoteServers,
+                (remoteServer) -> AtlasDbHttpClients.createProxy(optionalSecurity, remoteServer, ClockService.class));
+
+        Map<String, RequestTime> previousRequests = Maps.newHashMap();
+        for (String server : remoteServers) {
+            previousRequests.put(server, RequestTime.EMPTY);
+        }
+
+        return new ClockSkewMonitor(monitors, previousRequests, shouldRunClockSkewMonitor);
+    }
+
+    private ClockSkewMonitor(
+            Map<String, ClockService> monitorsByServer,
+            Map<String, RequestTime> previousRequestsByServer,
+            Supplier<Boolean> shouldRunClockSkewMonitor) {
+        this.monitorsByServer = monitorsByServer;
+        this.previousRequestsByServer = previousRequestsByServer;
+        this.shouldRunClockSkewMonitor = shouldRunClockSkewMonitor;
+    }
+
+    public void run() {
+        Executors.newSingleThreadScheduledExecutor().schedule(
+                () -> {
+                    try {
+                        if (shouldRunClockSkewMonitor.get()) {
+                            runInternal();
+                        }
+                    } catch (Throwable t) {
+                        log.warn("ClockSkewMonitor threw an exception", t);
+                    }
+                }, SLEEP_TIME_SECONDS, TimeUnit.SECONDS
+        );
+    }
+
+    private void runInternal() {
+        monitorsByServer.forEach((server, monitor) -> {
+            long localTimeAtStart = System.nanoTime();
+            long remoteSystemTime = monitor.getSystemTimeInNanos();
+            long localTimeAtEnd = System.nanoTime();
+
+            RequestTime previousRequest = previousRequestsByServer.get(server);
+            RequestTime newRequest = new RequestTime(localTimeAtStart, localTimeAtEnd, remoteSystemTime);
+            compareClockSkew(server, newRequest, previousRequest);
+        });
+    }
+
+    private void compareClockSkew(String server, RequestTime newRequest, RequestTime previousRequest) {
+        if (previousRequest.equals(RequestTime.EMPTY)) {
+            previousRequestsByServer.put(server, newRequest);
+            return;
+        }
+
+        long maxElapsedTime = newRequest.localTimeAtEnd - previousRequest.localTimeAtStart;
+        long minElapsedTime = newRequest.localTimeAtStart - previousRequest.localTimeAtEnd;
+        long remoteElapsedTime = newRequest.remoteSystemTime - previousRequest.remoteSystemTime;
```

Comment:
one pattern that might be useful here is to create a new class that holds the previous and new request as instance variables. then, the methods below could look like:
```
if (hasTooMuchTimeElapsedSincePreviousRequest()) {
   ...
}

if (requestTookTooLongToComplete()) {
   ...
}
```

etc..